In [ ]:
!pip install boto3
!pip install raceplotly

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 139.3/139.3 kB 3.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 11.8/11.8 MB 34.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 82.0/82.0 kB 10.6 MB/s eta 0:00:00


In [ ]:
import os

In [ ]:
os.environ["AWS_ACCESS_KEY_ID"] = 'AKIA2ADGJGHLWWVT2JUH'
os.environ["AWS_SECRET_ACCESS_KEY"] = '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'
os.environ["AWS_DEFAULT_REGION"] = 'us-west-2'

In [ ]:
import boto3
import pandas as pd
from io import StringIO

# Create an S3 client
s3 = boto3.client('s3')

In [ ]:
import boto3


def hello_s3():
    """
    Use the AWS SDK for Python (Boto3) to create an Amazon Simple Storage Service
    (Amazon S3) resource and list the buckets in your account.
    This example uses the default settings specified in your shared credentials
    and config files.
    """
    s3_resource = boto3.resource("s3", aws_access_key_id="AKIA2ADGJGHLWWVT2JUH",
                                 aws_secret_access_key="QkwrjwaLRUh1joFd0ZUG83AQmUW2+r8DtnKX6P4/")
    print("Hello, Amazon S3! Let's list your buckets:")
    for bucket in s3_resource.buckets.all():
        print(f"\t{bucket.name}")


if __name__ == "__main__":
    hello_s3()
# snippet-end:[python.example_code.s3.Hello]

Hello, Amazon S3! Let's list your buckets:
	mlb-data-clean
	mlb-data-store


In [ ]:
# Specify your bucket name and file name
bucket_name = 'mlb-data-clean'
file_name = 'mlb_digital_engagement_cleaned/mlb_data.csv'

# Get the object from S3
obj = s3.get_object(Bucket=bucket_name, Key=file_name)

In [ ]:
import pandas as pd

# Define the chunk size
chunk_size = 10000

# Create a list to hold chunks
chunks_list = []

# Read the file in chunks
for chunk in pd.read_csv(obj['Body'], chunksize=chunk_size):
    # Process each chunk (here, we simply add it to our list)
    chunks_list.append(chunk)

# Concatenate all chunks into a single DataFrame
df = pd.concat(chunks_list, axis=0)

# Now 'df' is a DataFrame containing all the data from the CSV file


In [ ]:
df.head()

date                            nextDayPlayerEngagement games  \
0  20180101  [{"engagementMetricsDate":"2018-01-02","player...   NaN   
1  20180102  [{"engagementMetricsDate":"2018-01-03","player...   NaN   
2  20180103  [{"engagementMetricsDate":"2018-01-04","player...   NaN   
3  20180104  [{"engagementMetricsDate":"2018-01-05","player...   NaN   
4  20180105  [{"engagementMetricsDate":"2018-01-06","player...   NaN   

                                             rosters playerBoxScores  \
0  [{"playerId":400121,"gameDate":"2018-01-01","t...             NaN   
1  [{"playerId":134181,"gameDate":"2018-01-02","t...             NaN   
2  [{"playerId":425492,"gameDate":"2018-01-03","t...             NaN   
3  [{"playerId":282332,"gameDate":"2018-01-04","t...             NaN   
4  [{"playerId":282332,"gameDate":"2018-01-05","t...             NaN   

  teamBoxScores                                       transactions standings  \
0           NaN  [{"transactionId":340732,"playerId":547348,"pl...       NaN   
1           NaN  [{"transactionId":339458,"playerId":621173,"pl...       NaN   
2           NaN  [{"transactionId":347527,"playerId":572389,"pl...       NaN   
3           NaN  [{"transactionId":339549,"playerId":545343,"pl...       NaN   
4           NaN  [{"transactionId":341195,"playerId":628336,"pl...       NaN   

  awards events                             playerTwitterFollowers  \
0    NaN    NaN  [{"date":"2018-01-01","playerId":545361,"playe...   
1    NaN    NaN                                                NaN   
2    NaN    NaN                                                NaN   
3    NaN    NaN                                                NaN   
4    NaN    NaN                                                NaN   

                                teamTwitterFollowers  
0  [{"date":"2018-01-01","teamId":147,"teamName":...  
1                                                NaN  
2                                                NaN  
3                                                NaN  
4                                                NaN

# **Baseball Analytics: A deep dive into MLB Data** #

## Datasets
* **mlb_data.csv** - Dataset at a playerId, date level. Contains nested JSONs for following - ```games, rosters, playerBoxScores, teamBoxScores, awards, events, playerEngagement etc.``` etc. The 4 playerEngagement variables are - ```target1, target2, target3, target4``` target1-target4 are each daily indexes of digital engagement on a 0-100 scale. Data is from 2018 - 2021.
* **teams.csv** - Dataset containing info for teams like ```teamName, location, league & division``` that they play in.
* **seasons.csv** - Contains dates for various phases of a season - ``preSeason, regularSeason, playoffs, postseason`` etc. from 2017-2021
* **players.csv** - Dataset containing info for players - ```DOB, birthcountry,mlbDebutDate, position``` etc.

## Initial Questions on Data<a name="que"></a>

* How do the four target engagement variables vary from one year to the other and how are they related to each other?
* How does engagament score decay with time after a baseball game?
* How are engagement score values autocorrelated with each other?
* How do engagement scores vary with time of the year & progress of the season? Is there seasonality associated with these scores?
* Does successful teams enjoy higher engagement scores compared to remaining teams? Do higher ranked teams see more engagement from fans?
* How does starting position of a player play a role in driving engagement scores? Are players in certain positions likely to drive higher engagement?
* How does a player's demographic attributes play a role in determining their engagement scores?
* How do run scoring & pitching stats impact engagement scores? Does good & bad performance both lead to higher engagement from fans?

## Importing the datasets<a name="import"></a>

The section of code below unpacks the JSON for various columns in the ```mlb_data.csv``` and creates a nice tabular dataset for us to run further analyses.




In [ ]:
# Importing necessary libraries and modules
import gc  # Garbage Collector interface (for memory management)
import sys  # System-specific parameters and functions
import warnings  # To control warning messages
from pathlib import Path  # Object-oriented filesystem paths

import ipywidgets as widgets  # Interactive widgets for Jupyter notebooks
import matplotlib.pyplot as plt  # Plotting library
import numpy as np  # Numerical operations and array handling
import pandas as pd  # Data manipulation and analysis
import seaborn as sns  # Statistical data visualization

from raceplotly.plots import barplot  # Barplot module from raceplotly for animated plots

warnings.simplefilter("ignore")  # Ignore warnings to clean up output

# Set Matplotlib defaults for plotting
plt.style.use("seaborn-whitegrid")  # Using seaborn style for matplotlib plots
plt.rc("figure", autolayout=True, figsize=(11, 5))  # Default figure size and autolayout for all plots
plt.rc(
    "axes",
    labelweight="bold",
    labelsize="large",
    titleweight="bold",
    titlesize=14,
    titlepad=10,
)  # Customizing the axes' appearance in all plots
plot_params = dict(
    color="0.75",
    style=".-",
    markeredgecolor="0.25",
    markerfacecolor="0.25",
    legend=False,
)  # Default plot parameters for consistent styling


# Load and Join Data #

In [ ]:
import pandas as pd
import boto3

def read_csv_from_s3(bucket_name, file_path, chunk_size=10000):
    # Create an S3 client
    s3 = boto3.client('s3')

    # Get the object from S3
    obj = s3.get_object(Bucket=bucket_name, Key=file_path)

    # Create a list to hold chunks
    chunks_list = []

    # Read the file in chunks
    for chunk in pd.read_csv(obj['Body'], chunksize=chunk_size):
        chunks_list.append(chunk)

    # Concatenate all chunks into a single DataFrame
    return pd.concat(chunks_list, axis=0)

# Usage
bucket_name = 'mlb-data-clean'
training = read_csv_from_s3(bucket_name, 'mlb_digital_engagement_cleaned/train.csv')
players = read_csv_from_s3(bucket_name, 'mlb_digital_engagement_cleaned/players.csv')

# Convert training data date field to datetime type
training['date'] = pd.to_datetime(training['date'], format="%Y%m%d")

# Display information
display(training.info())


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1216 entries, 0 to 1215
Data columns (total 12 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   date                     1216 non-null   datetime64[ns]
 1   nextDayPlayerEngagement  1216 non-null   object        
 2   games                    639 non-null    object        
 3   rosters                  1216 non-null   object        
 4   playerBoxScores          538 non-null    object        
 5   teamBoxScores            538 non-null    object        
 6   transactions             1103 non-null   object        
 7   standings                531 non-null    object        
 8   awards                   294 non-null    object        
 9   events                   536 non-null    object        
 10  playerTwitterFollowers   40 non-null     object        
 11  teamTwitterFollowers     40 non-null     object        
dtypes: datetime64[ns](1), object(11)
m

None

In [ ]:
import pandas as pd
import boto3

# Create an S3 client
s3 = boto3.client('s3')
bucket_name = 'mlb-data-clean'

df_names = ['seasons', 'teams', 'players', 'awards']
chunk_size = 10000

for name in df_names:
    file_path = f"mlb_digital_engagement_cleaned/{name}.csv"
    obj = s3.get_object(Bucket=bucket_name, Key=file_path)

    chunks_list = []
    for chunk in pd.read_csv(obj['Body'], chunksize=chunk_size):
        chunks_list.append(chunk)

    globals()[name] = pd.concat(chunks_list, axis=0)

In [ ]:
_data_tabs = widgets.Tab()
# Add Output widgets for each pandas DF as tabs' children
_data_tabs.children = list([widgets.Output() for df_name in df_names])

for index in range(0, len(df_names)):
    # Rename tab bar titles to df names
    _data_tabs.set_title(index, df_names[index])

    # Display corresponding table output for this tab name
    with _data_tabs.children[index]:
        display(eval(df_names[index]))

display(_data_tabs)

In [ ]:
# Helper function to unpack json found in daily data
def unpack_json(json_str):
    return np.nan if pd.isna(json_str) else pd.read_json(json_str)

### Unnest various nested data within training (daily) data ####
daily_data_unnested_dfs = pd.DataFrame(data = {
  'dfName': training.drop('date', axis = 1).columns.values.tolist()
  })

daily_data_unnested_dfs['df'] = [pd.DataFrame() for row in
  daily_data_unnested_dfs.iterrows()]

for df_index, df_row in daily_data_unnested_dfs.iterrows():
    nestedTableName = str(df_row['dfName'])

    date_nested_table = training[['date', nestedTableName]]

    date_nested_table = (date_nested_table[
      ~pd.isna(date_nested_table[nestedTableName])
      ].
      reset_index(drop = True)
      )

    daily_dfs_collection = []

    for date_index, date_row in date_nested_table.iterrows():
        daily_df = unpack_json(date_row[nestedTableName])

        daily_df['dailyDataDate'] = date_row['date']

        daily_dfs_collection = daily_dfs_collection + [daily_df]

    unnested_table = pd.concat(daily_dfs_collection,
      ignore_index = True).set_index('dailyDataDate').reset_index()

    # Creates 1 pandas df per unnested df from daily data read in, with same name
    globals()[df_row['dfName']] = unnested_table

    daily_data_unnested_dfs['df'][df_index] = unnested_table

del training
gc.collect()



#### Get some information on each date in daily data (using season dates of interest) ####
dates = pd.DataFrame(data =
  {'dailyDataDate': nextDayPlayerEngagement['dailyDataDate'].unique()})

dates['date'] = pd.to_datetime(dates['dailyDataDate'].astype(str))

dates['year'] = dates['date'].dt.year
dates['month'] = dates['date'].dt.month

dates_with_info = pd.merge(
  dates,
  seasons,
  left_on = 'year',
  right_on = 'seasonId'
  )

dates_with_info['inSeason'] = (
  dates_with_info['date'].between(
    dates_with_info['regularSeasonStartDate'],
    dates_with_info['postSeasonEndDate'],
    inclusive = True
    )
  )

dates_with_info['seasonPart'] = np.select(
  [
    dates_with_info['date'] < dates_with_info['preSeasonStartDate'],
    dates_with_info['date'] < dates_with_info['regularSeasonStartDate'],
    dates_with_info['date'] <= dates_with_info['lastDate1stHalf'],
    dates_with_info['date'] < dates_with_info['firstDate2ndHalf'],
    dates_with_info['date'] <= dates_with_info['regularSeasonEndDate'],
    dates_with_info['date'] < dates_with_info['postSeasonStartDate'],
    dates_with_info['date'] <= dates_with_info['postSeasonEndDate'],
    dates_with_info['date'] > dates_with_info['postSeasonEndDate']
  ],
  [
    'Offseason',
    'Preseason',
    'Reg Season 1st Half',
    'All-Star Break',
    'Reg Season 2nd Half',
    'Between Reg and Postseason',
    'Postseason',
    'Offseason'
  ],
  default = np.nan
  )

#### Add some pitching stats/pieces of info to player game level stats ####

player_game_stats = (playerBoxScores.copy().
  # Change team Id/name to reflect these come from player game, not roster
  rename(columns = {'teamId': 'gameTeamId', 'teamName': 'gameTeamName'})
  )

# Adds in field for innings pitched as fraction (better for aggregation)
player_game_stats['inningsPitchedAsFrac'] = np.where(
  pd.isna(player_game_stats['inningsPitched']),
  np.nan,
  np.floor(player_game_stats['inningsPitched']) +
    (player_game_stats['inningsPitched'] -
      np.floor(player_game_stats['inningsPitched'])) * 10/3
  )

# Add in Tom Tango pitching game score (https://www.mlb.com/glossary/advanced-stats/game-score)
player_game_stats['pitchingGameScore'] = (40
    + 2 * player_game_stats['outs']
    + 1 * player_game_stats['strikeOutsPitching']
    - 2 * player_game_stats['baseOnBallsPitching']
    - 2 * player_game_stats['hitsPitching']
    - 3 * player_game_stats['runsPitching']
    - 6 * player_game_stats['homeRunsPitching']
    )

# Add in criteria for no-hitter by pitcher (individual, not multiple pitchers)
player_game_stats['noHitter'] = np.where(
  (player_game_stats['gamesStartedPitching'] == 1) &
  (player_game_stats['inningsPitched'] >= 9) &
  (player_game_stats['hitsPitching'] == 0),
  1, 0
  )

player_date_stats_agg = pd.merge(
  (player_game_stats.
    groupby(['dailyDataDate', 'playerId'], as_index = False).
    # Some aggregations that are not simple sums
    agg(
      numGames = ('gamePk', 'nunique'),
      # Should be 1 team per player per day, but adding here for 1 exception:
      # playerId 518617 (Jake Diekman) had 2 games for different teams marked
      # as played on 5/19/19, due to resumption of game after he was traded
      numTeams = ('gameTeamId', 'nunique'),
      # Should be only 1 team for almost all player-dates, taking min to simplify
      gameTeamId = ('gameTeamId', 'min')
      )
    ),
  # Merge with a bunch of player stats that can be summed at date/player level
  (player_game_stats.
    groupby(['dailyDataDate', 'playerId'], as_index = False)
    [['runsScored', 'homeRuns', 'strikeOuts', 'baseOnBalls', 'hits',
      'hitByPitch', 'atBats', 'caughtStealing', 'stolenBases',
      'groundIntoDoublePlay', 'groundIntoTriplePlay', 'plateAppearances',
      'totalBases', 'rbi', 'leftOnBase', 'sacBunts', 'sacFlies',
      'gamesStartedPitching', 'runsPitching', 'homeRunsPitching',
      'strikeOutsPitching', 'baseOnBallsPitching', 'hitsPitching',
      'inningsPitchedAsFrac', 'earnedRuns',
      'battersFaced','saves', 'blownSaves', 'pitchingGameScore',
      'noHitter'
      ]].
    sum()
    ),
  on = ['dailyDataDate', 'playerId'],
  how = 'inner'
  )

#### Turn games table into 1 row per team-game, then merge with team box scores ####
# Filter to regular or Postseason games w/ valid scores for this part
games_for_stats = games[
  np.isin(games['gameType'], ['R', 'F', 'D', 'L', 'W', 'C', 'P']) &
  ~pd.isna(games['homeScore']) &
  ~pd.isna(games['awayScore'])
  ]

# Get games table from home team perspective
games_home_perspective = games_for_stats.copy()

# Change column names so that "team" is "home", "opp" is "away"
games_home_perspective.columns = [
  col_value.replace('home', 'team').replace('away', 'opp') for
    col_value in games_home_perspective.columns.values]

games_home_perspective['isHomeTeam'] = 1

# Get games table from away team perspective
games_away_perspective = games_for_stats.copy()

# Change column names so that "opp" is "home", "team" is "away"
games_away_perspective.columns = [
  col_value.replace('home', 'opp').replace('away', 'team') for
    col_value in games_away_perspective.columns.values]

games_away_perspective['isHomeTeam'] = 0

# Put together games from home/away perspective to get df w/ 1 row per team game
team_games = (pd.concat([
  games_home_perspective,
  games_away_perspective
  ],
  ignore_index = True)
  )

# Copy over team box scores data to modify
team_game_stats = teamBoxScores.copy()

# Change column names to reflect these are all "team" stats - helps
# to differentiate from individual player stats if/when joining later
team_game_stats.columns = [
  (col_value + 'Team')
  if (col_value not in ['dailyDataDate', 'home', 'teamId', 'gamePk',
    'gameDate', 'gameTimeUTC'])
    else col_value
  for col_value in team_game_stats.columns.values
  ]

# Merge games table with team game stats
team_games_with_stats = pd.merge(
  team_games,
  team_game_stats.
    # Drop some fields that are already present in team_games table
    drop(['home', 'gameDate', 'gameTimeUTC'], axis = 1),
  on = ['dailyDataDate', 'gamePk', 'teamId'],
  # Doing this as 'inner' join excludes spring training games, postponed games,
  # etc. from original games table, but this may be fine for purposes here
  how = 'inner'
  )

team_date_stats_agg = (team_games_with_stats.
  groupby(['dailyDataDate', 'teamId', 'gameType', 'oppId', 'oppName'],
    as_index = False).
  agg(
    numGamesTeam = ('gamePk', 'nunique'),
    winsTeam = ('teamWinner', 'sum'),
    lossesTeam = ('oppWinner', 'sum'),
    runsScoredTeam = ('teamScore', 'sum'),
    runsAllowedTeam = ('oppScore', 'sum')
    )
   )

# Prepare standings table for merge w/ player digital engagement data
# Pick only certain fields of interest from standings for merge
standings_selected_fields = (standings[['dailyDataDate', 'teamId',
  'streakCode', 'divisionRank', 'leagueRank', 'wildCardRank', 'pct'
  ]].
  rename(columns = {'pct': 'winPct'})
  )

# Change column names to reflect these are all "team" standings - helps
# to differentiate from player-related fields if/when joining later
standings_selected_fields.columns = [
  (col_value + 'Team')
  if (col_value not in ['dailyDataDate', 'teamId'])
    else col_value
  for col_value in standings_selected_fields.columns.values
  ]

standings_selected_fields['streakLengthTeam'] = (
  standings_selected_fields['streakCodeTeam'].
    str.replace('W', '').
    str.replace('L', '').
    astype(float)
    )

# Add fields to separate winning and losing streak from streak code
standings_selected_fields['winStreakTeam'] = np.where(
  standings_selected_fields['streakCodeTeam'].str[0] == 'W',
  standings_selected_fields['streakLengthTeam'],
  np.nan
  )

standings_selected_fields['lossStreakTeam'] = np.where(
  standings_selected_fields['streakCodeTeam'].str[0] == 'L',
  standings_selected_fields['streakLengthTeam'],
  np.nan
  )

standings_for_digital_engagement_merge = (pd.merge(
  standings_selected_fields,
  dates_with_info[['dailyDataDate', 'inSeason']],
  on = ['dailyDataDate'],
  how = 'left'
  ).
  # Limit down standings to only in season version
  query("inSeason").
  # Drop fields no longer necessary (in derived values, etc.)
  drop(['streakCodeTeam', 'streakLengthTeam', 'inSeason'], axis = 1).
  reset_index(drop = True)
  )

#### Merge together various data frames to add date, player, roster, and team info ####
# Copy over player engagement df to add various pieces to it
player_engagement_with_info = nextDayPlayerEngagement.copy()

# Take "row mean" across targets to add (helps with studying all 4 targets at once)
player_engagement_with_info['targetAvg'] = np.mean(
  player_engagement_with_info[['target1', 'target2', 'target3', 'target4']],
  axis = 1)

# Merge in date information
player_engagement_with_info = pd.merge(
  player_engagement_with_info,
  dates_with_info[['dailyDataDate', 'date', 'year', 'month', 'inSeason',
    'seasonPart']],
  on = ['dailyDataDate'],
  how = 'left'
  )

# Merge in some player information
player_engagement_with_info = pd.merge(
  player_engagement_with_info,
  players[['playerId', 'playerName', 'DOB', 'mlbDebutDate', 'birthCity',
    'birthStateProvince', 'birthCountry', 'primaryPositionName']],
   on = ['playerId'],
   how = 'left'
   )

# Merge in some player roster information by date
player_engagement_with_info = pd.merge(
  player_engagement_with_info,
  (rosters[['dailyDataDate', 'playerId', 'statusCode', 'status', 'teamId']].
    rename(columns = {
      'statusCode': 'rosterStatusCode',
      'status': 'rosterStatus',
      'teamId': 'rosterTeamId'
      })
    ),
  on = ['dailyDataDate', 'playerId'],
  how = 'left'
  )

# Merge in team name from player's roster team
player_engagement_with_info = pd.merge(
  player_engagement_with_info,
  (teams[['id', 'teamName']].
    rename(columns = {
      'id': 'rosterTeamId',
      'teamName': 'rosterTeamName'
      })
    ),
  on = ['rosterTeamId'],
  how = 'left'
  )

# Merge in some player game stats (previously aggregated) from that date
player_engagement_with_info = pd.merge(
  player_engagement_with_info,
  player_date_stats_agg,
  on = ['dailyDataDate', 'playerId'],
  how = 'left'
  )

# Merge in team name from player's game team
player_engagement_with_info = pd.merge(
  player_engagement_with_info,
  (teams[['id', 'teamName']].
    rename(columns = {
      'id': 'gameTeamId',
      'teamName': 'gameTeamName'
      })
    ),
  on = ['gameTeamId'],
  how = 'left'
  )

# Merge in some team game stats/results (previously aggregated) from that date
player_engagement_with_info = pd.merge(
  player_engagement_with_info,
  team_date_stats_agg.rename(columns = {'teamId': 'gameTeamId'}),
  on = ['dailyDataDate', 'gameTeamId'],
  how = 'left'
  )

# Merge in player transactions of note on that date

# Merge in some pieces of team standings (previously filter/processed) from that date
player_engagement_with_info = pd.merge(
  player_engagement_with_info,
  standings_for_digital_engagement_merge.
    rename(columns = {'teamId': 'gameTeamId'}),
  on = ['dailyDataDate', 'gameTeamId'],
  how = 'left'
  )

#display(player_engagement_with_info)
player_engagement_with_info.head(5)


## Exploratory Analysis<a name="exp"></a>

As you can see above, the code converts the initial ```mlb_data.csv``` dataset with 1216 rows & several packed JSONs into a dataset with 2.5M+ rows. The processed dataframe ```player_engagement_with_info``` contains data in a much more simpler format for data analysis. Let's start exploring the dataset and start answering the questions highlighted at the start of the notebook.



In [ ]:
import pickle

In [ ]:
player_engagement_with_info.to_pickle('player_engagement_with_info.pkl')

In [ ]:
import boto3

# Specify your bucket name and the path in the bucket
bucket_name = 'mlb-data-clean'
s3_folder = 'mlb_digital_engagement_cleaned/'
file_name = '/player_engagement_with_info.pkl'
key = s3_folder + 'player_engagaement_with_info.pkl'

# Create an S3 client
s3 = boto3.client('s3')

# Upload the file
s3.upload_file(file_name, bucket_name, key)


In [ ]:
import boto3
import pandas as pd
from io import BytesIO

# Specify your bucket name and the path in the bucket
bucket_name = 'mlb-data-clean'
s3_folder = 'mlb_digital_engagement_cleaned/'
key = s3_folder + 'player_engagaement_with_info.pkl'

# Create an S3 client
s3 = boto3.client('s3')

# Get the object from S3
obj = s3.get_object(Bucket=bucket_name, Key=key)
data = obj['Body'].read()

# Read the data from the buffer
player_engagement_with_info = pd.read_pickle(BytesIO(data))

# Now 'player_engagement_with_info' is a DataFrame containing our data


In [ ]:
player_engagement_with_info

dailyDataDate engagementMetricsDate  playerId   target1    target2  \
0          2018-01-01            2018-01-02    628317  0.011167   4.474708   
1          2018-01-01            2018-01-02    547989  0.042993   5.593385   
2          2018-01-01            2018-01-02    519317  0.974327  56.177043   
3          2018-01-01            2018-01-02    607625  0.006700   2.675097   
4          2018-01-01            2018-01-02    592547  0.001117   0.632296   
...               ...                   ...       ...       ...        ...   
2506171    2021-04-30            2021-05-01    451661  0.000000   0.013314   
2506172    2021-04-30            2021-05-01    519301  0.000131   0.003329   
2506173    2021-04-30            2021-05-01    527055  0.000000   0.019971   
2506174    2021-04-30            2021-05-01    543484  0.000131   0.056586   
2506175    2021-04-30            2021-05-01    605525  0.000131   0.499284   

           target3    target4  targetAvg       date  year  ...  winsTeam  \
0         0.005168   5.735294   2.556584 2018-01-01  2018  ...       NaN   
1         0.045033   2.794118   2.118882 2018-01-01  2018  ...       NaN   
2        13.693746  64.166667  33.752945 2018-01-01  2018  ...       NaN   
3         0.005168   1.862745   1.137428 2018-01-01  2018  ...       NaN   
4         0.002953   0.931373   0.391934 2018-01-01  2018  ...       NaN   
...            ...        ...        ...        ...   ...  ...       ...   
2506171   0.000000   0.625925   0.159810 2021-04-30  2021  ...       NaN   
2506172   0.000000   0.216229   0.054922 2021-04-30  2021  ...       NaN   
2506173   0.000000   0.273131   0.073276 2021-04-30  2021  ...       NaN   
2506174   0.000000   1.024240   0.270239 2021-04-30  2021  ...       NaN   
2506175   0.000000   0.386935   0.221588 2021-04-30  2021  ...       NaN   

         lossesTeam runsScoredTeam runsAllowedTeam divisionRankTeam  \
0               NaN            NaN             NaN              NaN   
1               NaN            NaN             NaN              NaN   
2               NaN            NaN             NaN              NaN   
3               NaN            NaN             NaN              NaN   
4               NaN            NaN             NaN              NaN   
...             ...            ...             ...              ...   
2506171         NaN            NaN             NaN              NaN   
2506172         NaN            NaN             NaN              NaN   
2506173         NaN            NaN             NaN              NaN   
2506174         NaN            NaN             NaN              NaN   
2506175         NaN            NaN             NaN              NaN   

        leagueRankTeam wildCardRankTeam winPctTeam winStreakTeam  \
0                  NaN              NaN        NaN           NaN   
1                  NaN              NaN        NaN           NaN   
2                  NaN              NaN        NaN           NaN   
3                  NaN              NaN        NaN           NaN   
4                  NaN              NaN        NaN           NaN   
...                ...              ...        ...           ...   
2506171            NaN              NaN        NaN           NaN   
2506172            NaN              NaN        NaN           NaN   
2506173            NaN              NaN        NaN           NaN   
2506174            NaN              NaN        NaN           NaN   
2506175            NaN              NaN        NaN           NaN   

        lossStreakTeam  
0                  NaN  
1                  NaN  
2                  NaN  
3                  NaN  
4                  NaN  
...                ...  
2506171            NaN  
2506172            NaN  
2506173            NaN  
2506174            NaN  
2506175            NaN  

[2506176 rows x 72 columns]

Some of the columns present in the dataset above are -

### Player Info -
* DOB
* MLB Debut Date
* Birth Country
* Position
* Roster Status

#### Player Batting Scores -
* Runs Scored
* Home Runs
* Strike Outs
* RBI - Runs Batted In

#### Player Pitching Scores -
* Runs Pitching
* Strike Outs Pitching
* Hits Pitching

#### Team Stats -
* Team & Opposition Info
* Runs Scored & Allowed
* Wins & Loss Record

#### Target Columns - Target1 - Target 4

Let's first start by taking a look at Target Columns -

### Target Columns


In [ ]:
# Importing the Plotly Express module for plotting
import plotly.express as px

# Grouping the player engagement data by date and calculating the mean of each target variable
time_target_gp = player_engagement_with_info.groupby(['dailyDataDate']).agg({
    'target1': 'mean',
    'target2': 'mean',
    'target3': 'mean',
    'target4': 'mean',
    'targetAvg': 'mean'
}).reset_index()

# Importing necessary modules for creating subplots
from plotly.subplots import make_subplots
import plotly.graph_objects as go

# Creating a subplot grid of 3 rows and 2 columns
fig = make_subplots(rows=3, cols=2)

# Adding a scatter plot for target1 in the first row, first column
fig.add_trace(
    go.Scatter(x=time_target_gp['dailyDataDate'], y=time_target_gp['target1'], name='Target1'),
    row=1, col=1
)

# Adding a scatter plot for target2 in the first row, second column
fig.add_trace(
    go.Scatter(x=time_target_gp['dailyDataDate'], y=time_target_gp['target2'], name='Target2'),
    row=1, col=2
)

# Adding a scatter plot for target3 in the second row, first column
fig.add_trace(
    go.Scatter(x=time_target_gp['dailyDataDate'], y=time_target_gp['target3'], name='Target3'),
    row=2, col=1
)

# Adding a scatter plot for target4 in the second row, second column
fig.add_trace(
    go.Scatter(x=time_target_gp['dailyDataDate'], y=time_target_gp['target4'], name='Target4'),
    row=2, col=2
)

# Adding a scatter plot for the average target in the third row, first column
fig.add_trace(
    go.Scatter(x=time_target_gp['dailyDataDate'], y=time_target_gp['targetAvg'], name='targetAvg'),
    row=3, col=1
)

# Updating the layout of the figure with a specified height, width, and title
fig.update_layout(height=600, width=800, title_text="Daily Average Target Variable Values")

# Displaying the figure
fig.show()


In [ ]:
player_engagement_with_info['quarter']=player_engagement_with_info['dailyDataDate'].dt.quarter
player_engagement_with_info['month']=player_engagement_with_info['dailyDataDate'].dt.month
player_engagement_with_info['year']=player_engagement_with_info['dailyDataDate'].dt.year

player_engagement_with_info

dailyDataDate engagementMetricsDate  playerId   target1    target2  \
0          2018-01-01            2018-01-02    628317  0.011167   4.474708   
1          2018-01-01            2018-01-02    547989  0.042993   5.593385   
2          2018-01-01            2018-01-02    519317  0.974327  56.177043   
3          2018-01-01            2018-01-02    607625  0.006700   2.675097   
4          2018-01-01            2018-01-02    592547  0.001117   0.632296   
...               ...                   ...       ...       ...        ...   
2506171    2021-04-30            2021-05-01    451661  0.000000   0.013314   
2506172    2021-04-30            2021-05-01    519301  0.000131   0.003329   
2506173    2021-04-30            2021-05-01    527055  0.000000   0.019971   
2506174    2021-04-30            2021-05-01    543484  0.000131   0.056586   
2506175    2021-04-30            2021-05-01    605525  0.000131   0.499284   

           target3    target4  targetAvg       date  year  ...  lossesTeam  \
0         0.005168   5.735294   2.556584 2018-01-01  2018  ...         NaN   
1         0.045033   2.794118   2.118882 2018-01-01  2018  ...         NaN   
2        13.693746  64.166667  33.752945 2018-01-01  2018  ...         NaN   
3         0.005168   1.862745   1.137428 2018-01-01  2018  ...         NaN   
4         0.002953   0.931373   0.391934 2018-01-01  2018  ...         NaN   
...            ...        ...        ...        ...   ...  ...         ...   
2506171   0.000000   0.625925   0.159810 2021-04-30  2021  ...         NaN   
2506172   0.000000   0.216229   0.054922 2021-04-30  2021  ...         NaN   
2506173   0.000000   0.273131   0.073276 2021-04-30  2021  ...         NaN   
2506174   0.000000   1.024240   0.270239 2021-04-30  2021  ...         NaN   
2506175   0.000000   0.386935   0.221588 2021-04-30  2021  ...         NaN   

         runsScoredTeam runsAllowedTeam divisionRankTeam leagueRankTeam  \
0                   NaN             NaN              NaN            NaN   
1                   NaN             NaN              NaN            NaN   
2                   NaN             NaN              NaN            NaN   
3                   NaN             NaN              NaN            NaN   
4                   NaN             NaN              NaN            NaN   
...                 ...             ...              ...            ...   
2506171             NaN             NaN              NaN            NaN   
2506172             NaN             NaN              NaN            NaN   
2506173             NaN             NaN              NaN            NaN   
2506174             NaN             NaN              NaN            NaN   
2506175             NaN             NaN              NaN            NaN   

        wildCardRankTeam winPctTeam winStreakTeam lossStreakTeam quarter  
0                    NaN        NaN           NaN            NaN       1  
1                    NaN        NaN           NaN            NaN       1  
2                    NaN        NaN           NaN            NaN       1  
3                    NaN        NaN           NaN            NaN       1  
4                    NaN        NaN           NaN            NaN       1  
...                  ...        ...           ...            ...     ...  
2506171              NaN        NaN           NaN            NaN       2  
2506172              NaN        NaN           NaN            NaN       2  
2506173              NaN        NaN           NaN            NaN       2  
2506174              NaN        NaN           NaN            NaN       2  
2506175              NaN        NaN           NaN            NaN       2  

[2506176 rows x 73 columns]

In [ ]:
# Group the player engagement data by month and calculate the mean of each target variable
month_gp = player_engagement_with_info.groupby('month').agg({
    'target1': 'mean',
    'target2': 'mean',
    'target3': 'mean',
    'target4': 'mean'
}).reset_index()

# Create an empty figure for plotting
fig = go.Figure()

# Add a line plot for target1
fig.add_trace(go.Scatter(x=month_gp['month'], y=month_gp['target1'],
                    mode='lines',
                    name='target1'))

# Add a line plot for target2
fig.add_trace(go.Scatter(x=month_gp['month'], y=month_gp['target2'],
                    mode='lines',
                    name='target2'))

# Add a line plot for target3
fig.add_trace(go.Scatter(x=month_gp['month'], y=month_gp['target3'],
                    mode='lines',
                    name='target3'))

# Add a line plot for target4
fig.add_trace(go.Scatter(x=month_gp['month'], y=month_gp['target4'],
                    mode='lines',
                    name='target4'))

# Update the layout of the figure with a title
fig.update_layout(title='Variation in Average Target Values with Month')

# Display the figure
fig.show()


In [ ]:
#Time Series Correlation
import seaborn as sns
corr=time_target_gp.corr()
corr

target1   target2   target3   target4  targetAvg
target1    1.000000  0.247508  0.352359  0.325641   0.479742
target2    0.247508  1.000000  0.323602  0.344020   0.890217
target3    0.352359  0.323602  1.000000  0.328463   0.547052
target4    0.325641  0.344020  0.328463  1.000000   0.679669
targetAvg  0.479742  0.890217  0.547052  0.679669   1.000000

In [ ]:
# Grouping the player engagement data by date and calculating the mean of each target variable
daily = player_engagement_with_info.groupby('dailyDataDate').agg({
    'target1': 'mean',
    'target2': 'mean',
    'target3': 'mean',
    'target4': 'mean'
}).reset_index()

# Creating a subplot grid of 3 rows and 3 columns with shared Y-axes
fig = make_subplots(
    rows=3, cols=3, shared_yaxes=True,
    subplot_titles=("Target1 vs Target2", "Target1 vs Target3", "Target1 vs Target4",
                    "Target2 vs Target3", "Target2 vs Target4", "",
                    "Target3 vs Target4"))

# Adding scatter plots for each pair of targets to the subplots
fig.add_trace(
    go.Scatter(y=daily['target2'], x=daily['target1'], mode='markers'),
    row=1, col=1
)
fig.add_trace(
    go.Scatter(y=daily['target3'], x=daily['target1'], mode='markers'),
    row=1, col=2
)
fig.add_trace(
    go.Scatter(y=daily['target4'], x=daily['target1'], mode='markers'),
    row=1, col=3
)
fig.add_trace(
    go.Scatter(y=daily['target3'], x=daily['target2'], mode='markers'),
    row=2, col=1
)
fig.add_trace(
    go.Scatter(y=daily['target4'], x=daily['target2'], mode='markers'),
    row=2, col=2
)
fig.add_trace(
    go.Scatter(y=daily['target4'], x=daily['target3'], mode='markers'),
    row=3, col=1
)

# Updating layout to hide the legend
fig.update_layout(showlegend=False)

# Displaying the figure
fig.show()


In [ ]:
player_mean=player_engagement_with_info.groupby('playerId').agg({'target1':'mean','target2':'mean','target3':'mean','target4':'mean'}).reset_index()
player_mean


fig=go.Figure()
# fig.add_trace(go.Histogram(x=player_mean['target1'],histnorm='probability',name='Target1'))
# fig.add_trace(go.Histogram(x=player_mean['target2'],histnorm='probability',name='Target2'))
# fig.add_trace(go.Histogram(x=player_mean['target3'],histnorm='probability',name='Target3'))
# fig.add_trace(go.Histogram(x=player_mean['target4'],histnorm='probability',name='Target4'))
# fig.update_layout(barmode='overlay',title='Distribution of Average Player Target Scores')
# fig.update_traces(opacity=0.75)

fig.add_trace(go.Violin(y=player_mean['target1'],name='Target1',box_visible=True,meanline_visible=True))
fig.add_trace(go.Violin(y=player_mean['target2'],name='Target2',box_visible=True,meanline_visible=True))
fig.add_trace(go.Violin(y=player_mean['target3'],name='Target3',box_visible=True,meanline_visible=True))
fig.add_trace(go.Violin(y=player_mean['target4'],name='Target4',box_visible=True,meanline_visible=True))
fig.update_layout(title='Distribution of Average Player Target Scores')
# fig.update_traces(opacity=0.75)

fig.show()



Some of the things that are visible from the charts above -
* Target variables seem to be cyclical in nature with same pattern repeating every year
* **The Target variables don't seem to have significant Pearson's correlation with each other**. However some combinations do seem to be related to each other - for ex. Target 2 doesn't have any relation with Target 3 but seems to vary along with Target 4
* Monthwise plot for the average of Target variables shows significant seasonality. **Target Values seem to peak in the month of March & then dip during the month of October**. Similar trend is observable in all the target variables
* The mean target values for Players above shows that the **mean value for Target2 is much higher than other target variables**.
* Each of these target variables have a high number of outlier values - showing that a smaller group of players tend to outperform the vast majority

If we also look at the sample records for the ```player_engagement_with_info``` table we can see a lot of blank values for game stats. It is because we are required to predict engagement for players even on days that they are not playing. Therefore on non game days these values will appear as blank. Let's take a look at the breakdown of gamedays vs non gamedays in the dataset and also the average values for engagement on game days and non game days -

In [ ]:
#create a flag for gamedays and non-gamedays for players
player_engagement_with_info['gameday']=~player_engagement_with_info.gameTeamId.isna()
player_engagement_with_info['flag']=0
player_engagement_with_info.loc[player_engagement_with_info.gameTeamId.isna(),'flag']=-1

player_engagement_with_info.gameday.value_counts()


False    2323600
True      182576
Name: gameday, dtype: int64

In [ ]:
gameday_gp=player_engagement_with_info.groupby(['gameday']).agg({'targetAvg':'mean'}).reset_index()
px.bar(gameday_gp,x='gameday',y='targetAvg',title='Difference between targetAvg on MatchDays vs non MatchDays')

As we can see above, the number of non gameday records in the dataset are almost 10x higher than gameday records. Also as expected, the engagement scores are much higher on game days compared to non gamedays. Let's now take a deeper look at the data to try and understand how does engagement score decay on non-playing days.


In [ ]:
player_engagement_with_info.sort_values(by=['playerId','dailyDataDate'],inplace=True,ascending=True)
player_engagement_with_info=player_engagement_with_info.reset_index(drop=True)
#player_engagement_with_info.loc[player_engagement_with_info['gameday']==True,'daysSinceLastGame']=0

player_engagement_with_info['daysSinceLastGame']=0

# chk=player_engagement_with_info.head(10000)
# chk.to_csv('chk.csv')

In [ ]:
player_engagement_with_info.sort_values(by=['playerId','dailyDataDate'],inplace=True,ascending=True)

In [ ]:
def count_consecutive_items_n_cols(df, col_name_list, output_col):
    # This function counts consecutive occurrences of items across multiple columns

    # cum_sum_list will store cumulative sums for each column in col_name_list
    # The cumsum is reset to 1 whenever a change in value is detected in the column
    cum_sum_list = [
        (df[col_name] != df[col_name].shift(1)).cumsum().tolist() for col_name in col_name_list
    ]

    # Group by a composite key made by combining values of cum_sum_list and
    # count the number of consecutive occurrences, then add 1 for the current row
    df[output_col] = df.groupby(
        ["_".join(map(str, x)) for x in zip(*cum_sum_list)]
    ).cumcount() + 1

    return df

# Apply the function to player_engagement_with_info to calculate 'daysSinceLastGame'
# This will count consecutive days based on 'playerId' and 'gameday' columns
player_engagement_with_info = count_consecutive_items_n_cols(
    player_engagement_with_info,
    ['playerId', 'gameday'],
    'daysSinceLastGame'
)

# For rows where 'gameday' is True, reset 'daysSinceLastGame' to 0
# This indicates the count should restart as this is a game day
player_engagement_with_info.loc[player_engagement_with_info['gameday'] == True, 'daysSinceLastGame'] = 0


In [ ]:
# Calculate Days Since Last Game
# Grouping the data by 'daysSinceLastGame' and aggregating average target and unique player count
lag_gp = player_engagement_with_info.groupby(['daysSinceLastGame']).agg({
    'targetAvg': 'mean',
    'playerId': 'nunique'
}).reset_index()

# Create a bar plot with days since last game on x-axis and average target on y-axis
px.bar(lag_gp, x='daysSinceLastGame', y='targetAvg', title='Variation in targetAvg with daysSinceLastGame')

# Creating a subplot with secondary y-axis
fig1 = make_subplots(specs=[[{"secondary_y": True}]])

# Adding a bar trace for average target
fig1.add_trace(
    go.Bar(x=lag_gp['daysSinceLastGame'], y=lag_gp['targetAvg'], name="targetAvg"),
    secondary_y=True,
)

# Adding a scatter trace for number of players
fig1.add_trace(
    go.Scatter(x=lag_gp['daysSinceLastGame'], y=lag_gp['playerId'], name="No. of Players"),
    secondary_y=False,
)

# Updating layout with title and x-axis label
fig1.update_layout(title='Variation in targetAvg by daysSinceLastGame', xaxis_title='daysSinceLastGame')

# Displaying the figure
fig1.show()

# Filtering the data for immediate drop in target average within 10 days after a game
immediate_drop = player_engagement_with_info.loc[player_engagement_with_info['daysSinceLastGame'] <= 10]

# Grouping the filtered data by 'daysSinceLastGame' and aggregating average target and unique player count
lag_gp = immediate_drop.groupby(['daysSinceLastGame']).agg({
    'targetAvg': 'mean',
    'playerId': 'nunique'
}).reset_index()

# Create a line plot with days since last game on x-axis and average target on y-axis
# This plot shows the drop in targetAvg in the immediate days after a game
fig2 = px.line(lag_gp, x='daysSinceLastGame', y='targetAvg', title='Drop in targetAvg in immediate days after a game')

# Displaying the figure
fig2.show()


In [ ]:
high_target=player_engagement_with_info.loc[player_engagement_with_info['daysSinceLastGame']>=1000]
high_target.loc[high_target.rosterTeamName.isna(),'rosterTeamName']='No Team'

high_target_players=high_target.groupby(['playerName']).agg({'targetAvg':'mean'}).reset_index()

high_target_players=high_target_players.sort_values(by='targetAvg',ascending=False).head(10)
px.bar(high_target_players,x='playerName',y='targetAvg',title='Non-Active players with highest targetAvg')


We calculated a column ```daysSinceLastGame``` that keeps a track of number of days since player last played a game. Some of these players have been inactive for upto 2-3 years before they even played a game. We can see the following things above -
* The target engagement is higher immediately on game day but then significantly declines with time
* The drop in target engagement is not steady. There are several peaks observable around 200, 800 and 1000 day period.
* We looked at the drop in engagement immediately after the game and observed it initially goes down but then picks up again around 3-5 day period and then goes down again
* The rise in target engagement for players after remaining inactive for 1000 days was a bit perplexing. Therefore we filtered the data for these players and identified such players. 22 year old Brailyn Marquez & 23 year old Alex Kiriloff had the highest engagement scores
* A quick Google search revealed that some of the reasons for high engagement are player transfers & players returning from injury after a long break, leading to an increase in engagement.

In [ ]:
avg_gp_between_games=immediate_drop.groupby(['daysSinceLastGame']).agg({'playerId':'count'}).reset_index()
avg_gp_between_games

daysSinceLastGame  playerId
0                   0    182576
1                   1     87122
2                   2     51670
3                   3     36524
4                   4     28925
5                   5     19753
6                   6     13299
7                   7     11204
8                   8     10341
9                   9      9833
10                 10      9512

In [ ]:
team_gp=player_engagement_with_info.groupby('year').rosterTeamName.nunique()
team_gp

year
2018    30
2019    30
2020    30
2021    30
Name: rosterTeamName, dtype: int64

### Top Players

Let's try to understand who the most successful players are in the MLB over the years 2018-2021 and which positions they occupy on the field.

In [ ]:
player_high=player_engagement_with_info.groupby('playerId').agg({'targetAvg':'mean'}).reset_index()
player_high.sort_values(by='targetAvg',inplace=True,ascending=False)
player_high3=player_high.head(3)
player_high3

top3_players=player_engagement_with_info.loc[player_engagement_with_info['playerId'].isin(player_high3.playerId)]
top3_players=top3_players.groupby(['playerId','year','month']).agg({'targetAvg':'mean'}).reset_index()
top3_players['month_year'] = top3_players.year.astype(str) + '_'+top3_players.month.astype(str)

players_info=players[['playerId','playerName','primaryPositionName']]

top3_players=pd.merge(left=top3_players,right=players_info,left_on='playerId',right_on='playerId',how='inner')


px.line(top3_players,x='month_year',y='targetAvg',color='playerName',title='targetAvg Engagement patterns for Top 3 Players')

In [ ]:
player_high100=player_high.head(100)

top100_players=pd.merge(left=player_high100,right=players_info,left_on='playerId',right_on='playerId',how='inner')
top100_players

fig=px.scatter(top100_players,y='targetAvg',color='primaryPositionName',title='targetAvg scores for top 100 players')
fig.show()

fig1 = px.pie(top100_players, values='targetAvg', names='primaryPositionName', title='Breakdown of top 100 players by position')

fig1.show()

Findings from the charts above -
* **Mike Trout, Bryce Harper & Aaron Judge stand out as the top 3 players during the period of 2018-2021**. Each of them had a really outstanding year where their targetAvg values was much higher than other years.
* As we noticed in the overall monthly targetAvg values chart, the targetAvg values for these top players stayed highest in March & then dipped to its lowest in October.
* I also took a look at the top 100 players with highest targetAvg values. **Top 4 players all turn out to be Designated Hitters**. Designated Hitter is a player who bats in place of a pitcher. The targetAvg values for these top 4 players are significantly higher than rest of the group
* I also looked at the breakdown of top 100 players by position. **30% of these players were Designated Hitters & approx another 30% were pitchers**, which shows that these players are much likelier to score higher than rest of the team players on engagement score metric.

Having taken a look at the top 100 players, let's take a look at all the players in the dataset and average scores by position. The image below shows positioning of different players on the baseball field -  

![](https://upload.wikimedia.org/wikipedia/commons/thumb/8/88/Baseball_positions.svg/1200px-Baseball_positions.svg.png)

In [ ]:
position_avg=player_engagement_with_info.groupby(['primaryPositionName']).agg({'targetAvg':'mean','playerId':'nunique'}).reset_index()
position_avg.sort_values(by='targetAvg',inplace=True,ascending=False)

fig=px.bar(position_avg,x='primaryPositionName',y='targetAvg',title='Target Avg values by Position')
fig.show()
position_avg.sort_values(by='playerId',inplace=True,ascending=False)

fig1=px.bar(position_avg,x='primaryPositionName',y='playerId',title='Count of Players by Position')
fig1.show()


player_high_join=pd.merge(left=player_high,right=players,how='inner')
#player_high_join=player_high_join.loc[player_high_join['primaryPositionName']=='Pitcher']
player_high_join

px.violin(player_high_join,y='targetAvg',box=True,x='primaryPositionName',title='Distribution of targetAvg by primaryPosition')

Here again we see that the average target scores for **Designated Hitters are higher than rest of the positions**. One thing though that comes across as a surprise is that average scores for pitchers is quite low and much behind several other positions. This shows that **while some pitchers are likely to be among the best, there are several pitchers who don't do so well over the season, bringing down the average of the entire group**.

Also we see that the overall distribution of players in the dataset seems to be heavily skewed towards pitchers, which does look a bit odd. The violin plot above shows that though there are lot of outliers in the pitchers group, the overall average for the group is quite low.

### Player Demographics



In [ ]:
player_data=player_engagement_with_info[['playerId','dailyDataDate','targetAvg']]
#player_data

player_demog_data=pd.merge(left=player_data,right=players,left_on='playerId',right_on='playerId',how='inner')

player_demog_data['DOB'] = pd.to_datetime(player_demog_data['DOB'], format="%Y-%m-%d")
player_demog_data['mlbDebutDate'] = pd.to_datetime(player_demog_data['mlbDebutDate'], format="%Y-%m-%d")

player_demog_data['age'] = (player_demog_data['dailyDataDate'] - player_demog_data['DOB']).dt.days
player_demog_data['age']=(player_demog_data['age']/365).astype(int)

player_demog_data['yearsSinceDebut'] = (player_demog_data['dailyDataDate'] - player_demog_data['mlbDebutDate']).dt.days
player_demog_data['yearsSinceDebut']=player_demog_data['yearsSinceDebut']/365
player_demog_data['yearsSinceDebut']=player_demog_data['yearsSinceDebut'].fillna(-1)
player_demog_data['yearsSinceDebut']=player_demog_data['yearsSinceDebut'].astype(int)
#player_demog_data

def plot_viz(colnames):
    j=1
    k=1
    leng=len(colnames)
    tup=('yearsSinceDebut vs targetAvg','Country of Birth vs targetAvg','Age vs targetAvg','Weight vs targetAvg')
    fig = make_subplots(rows=2, cols=2,subplot_titles=tup)
    for i in range(leng):
        gp=player_demog_data.groupby(colnames[i]).agg({'targetAvg':'mean'}).reset_index()
      #  print(i,j,k)
        fig.add_trace(go.Bar(x=gp.iloc[:,0],y=gp.iloc[:,1]),row=j,col=k)

        if(i==1):
            k=k+1
        if(j==2):
            j=1
        else :
            j=j+1
    fig.update_layout(showlegend=False)
    fig.show()
     #   fig.show()


colnames=['yearsSinceDebut','age','birthCountry','weight']
plot_viz(colnames)


#gp=player_demog_data.groupby(colnames[0]).agg({'targetAvg':'mean'}).reset_index()
#gp.iloc[:,0]

* Players performance generally tends to get better with age in MLB, which is a bit odd when you compare this to other sports. Players above 40 have a very high engagement scores, it could also be due to a small number of players in the group driving the average up
* New & young players generally don't do as well compared to their new counterparts. The average engagement scores tend to increase with levels of experience among players.
* Overseas players generally tend to better than American players on average. This could be due to the fact that as an overseas player, someone has to be really good to become a pro MLB player, whereas in USA it might be a bit easier to get notived and also due to the high volume of players the average gets lowered.
* There is no clear relation between player's weight & their targetAvg score performance


### Team Performances
Let's now look at the average scores for the teams involved over the years.

In [ ]:
team_gp=player_engagement_with_info.groupby(['rosterTeamName','year']).agg({'targetAvg':'mean','winsTeam':'sum','numGamesTeam':'sum'}).reset_index()
team_gp['pct_wins']=team_gp['winsTeam']/team_gp['numGamesTeam']
team_gp['year'] = team_gp['year'].apply(str)

team_gp.sort_values(by='targetAvg',inplace=True,ascending=False)
team_gp["team_year"] = team_gp["rosterTeamName"].astype(str) + '_'+team_gp["year"].astype(str)

team_gp

rosterTeamName  year  targetAvg  winsTeam  numGamesTeam  pct_wins  \
116        Yankees  2018   5.028703    1440.0        2390.0  0.602510   
117        Yankees  2019   4.330995    1630.0        2551.0  0.638965   
119        Yankees  2021   4.185222     172.0         370.0  0.464865   
38         Dodgers  2020   3.866232     875.0        1219.0  0.717801   
118        Yankees  2020   3.787553     582.0        1060.0  0.549057   
..             ...   ...        ...       ...           ...       ...   
101         Royals  2019   0.957269     840.0        2304.0  0.364583   
100         Royals  2018   0.938123     786.0        2214.0  0.355014   
52         Marlins  2018   0.901664     953.0        2456.0  0.388029   
68          Padres  2018   0.885004     980.0        2391.0  0.409870   
53         Marlins  2019   0.873908     834.0        2410.0  0.346058   

        team_year  
116  Yankees_2018  
117  Yankees_2019  
119  Yankees_2021  
38   Dodgers_2020  
118  Yankees_2020  
..            ...  
101   Royals_2019  
100   Royals_2018  
52   Marlins_2018  
68    Padres_2018  
53   Marlins_2019  

[120 rows x 7 columns]

In [ ]:
px.scatter(team_gp,x='pct_wins',y='targetAvg',text='team_year',title='Variation of targetAvg with Team WinPercent over Seasons')

In [ ]:
team_month_yr_gp=player_engagement_with_info.groupby(['rosterTeamName','year','month']).agg({'targetAvg':'mean'}).reset_index()
team_month_yr_gp=team_month_yr_gp.sort_values(by=['year','month'],ascending=True)
team_month_yr_gp['month_yr']=team_month_yr_gp['month'].astype(str)+"_"+team_month_yr_gp['year'].astype(str)
team_month_yr_gp['month_yr1'] = pd.to_datetime(team_month_yr_gp[['year', 'month']].assign(DAY=1))

team_month_yr_gp

# my_raceplot = barplot(team_month_yr_gp,  item_column='rosterTeamName', value_column='targetAvg', time_column='month_yr')
# my_raceplot.plot(item_label = 'team name', value_label = 'number of followers', frame_duration = 800,title="Top 10 teams with highest targetAvg score from 2018-2021")


rosterTeamName  year  month  targetAvg month_yr  month_yr1
0            Angels  2018      1   2.120307   1_2018 2018-01-01
40           Astros  2018      1   2.692390   1_2018 2018-01-01
80        Athletics  2018      1   0.909694   1_2018 2018-01-01
120       Blue Jays  2018      1   1.943814   1_2018 2018-01-01
160          Braves  2018      1   1.371934   1_2018 2018-01-01
...             ...   ...    ...        ...      ...        ...
1039         Royals  2021      4   1.330866   4_2021 2021-04-01
1079         Tigers  2021      4   1.672391   4_2021 2021-04-01
1119          Twins  2021      4   1.842185   4_2021 2021-04-01
1159      White Sox  2021      4   2.459547   4_2021 2021-04-01
1199        Yankees  2021      4   4.334693   4_2021 2021-04-01

[1200 rows x 6 columns]

In [ ]:
my_raceplot = barplot(team_month_yr_gp,  item_column='rosterTeamName', value_column='targetAvg', time_column='month_yr1')
my_raceplot.plot(item_label = 'team name', value_label = 'targetAvgScore', frame_duration = 800,title="Top 10 teams with highest targetAvg score from 2018-2021")


In [ ]:
rankscores=player_engagement_with_info.groupby('divisionRankTeam').agg({'targetAvg':'mean'}).reset_index()
rankscores

leaguerankscores=player_engagement_with_info.groupby('leagueRankTeam').agg({'targetAvg':'mean'}).reset_index()
leaguerankscores

px.line(leaguerankscores,x='leagueRankTeam',y='targetAvg',title='Variation in TargetAvg with League Rank for Team')

In [ ]:
team_player_gp=player_engagement_with_info.groupby(['rosterTeamName','playerName']).agg({'targetAvg':'mean'}).reset_index()
#team_player_gp.sort_values(['rosterTeamName','targetAvg'],ascending=['True','False'], inplace=True)

team_player_sum=team_player_gp.groupby('rosterTeamName').agg({'targetAvg':'sum'}).reset_index()
team_player_sum.columns=['rosterTeamName','targetAvgSum']

team_player_gp["rank"] = team_player_gp.groupby(['rosterTeamName'])["targetAvg"].rank("dense", ascending=False)
team_player_gp

team_player_gp_rank=pd.merge(left=team_player_gp,right=team_player_sum,how='inner')
team_player_gp_rank10=team_player_gp_rank.loc[team_player_gp_rank['rank']<=10]
team_player_gp_rank10.sort_values(by=['rosterTeamName','rank'],inplace=True)

team_player_gp_rank10['top10score']=team_player_gp_rank10.groupby(['rosterTeamName'])['targetAvg'].apply(lambda x: x.cumsum())
team_player_gp_rank10

team_player_gp_rank10_contribution=team_player_gp_rank10.loc[team_player_gp_rank10['rank']==10]
team_player_gp_rank10_contribution
team_player_gp_rank10_contribution['pct_contribution_top10']=team_player_gp_rank10_contribution['top10score']/team_player_gp_rank10_contribution['targetAvgSum']
team_player_gp_rank10_contribution.sort_values(by='pct_contribution_top10',ascending=False,inplace=True)

team_avg=player_engagement_with_info.groupby(['rosterTeamName']).agg({'targetAvg':'mean'}).reset_index()


team_player_gp_rank10_contribution=pd.merge(left=team_player_gp_rank10_contribution,right=team_avg,how='inner',on='rosterTeamName')

px.scatter(team_player_gp_rank10_contribution,y='pct_contribution_top10',x='targetAvgSum',size='targetAvg_y',text='rosterTeamName',title='Contribution of Top10 Players towards sum of team targetAvg scores' )


* Each point in the first scatter plot shows average values for target scores during a particular season. **Familiar team names like Yankees, Red Sox & Dodgers show up at the top, which shows that teams that have a high win percentage generally tend to have higher targetAvg as well**
* Yankees in 2018 has been the most engaged team
* Looking at the raceplot above, the list of top 10 teams has been fairly constant over time
* I also looked at the variation of targetAvg scores by leagueRankTeam. **Teams that have a better rank tend to have higher targetAvg scores as well.**
* We then looked at contribution of top 10 players towards the sum of targetAvg of all players in the team. Yankees clearly stand out here - they have the highest sum of targetAvg by a fair margin and only 30% of their total comes in from top10 players
* Less successful teams like Astros, Cardinals & Rockies are much more dependent on their top 10 players

Let's also quickly take a look at variation of targetAvg scores by game types as well

In [ ]:
gametype=player_engagement_with_info.groupby('gameType').agg({'targetAvg':'mean'}).reset_index()
px.bar(gametype,x='gameType',y='targetAvg',title='Variation of targetAvg with GameType')

In [ ]:
seasontype=player_engagement_with_info.groupby('seasonPart').agg({'targetAvg':'mean'}).reset_index()
px.bar(seasontype,x='seasonPart',y='targetAvg',title='Variation of targetAvg with seasonPart')

Preseason period tends to have the highest engagement score & postseason has the lowest targetAvg. The targetAvg scores tends to remain high during the Regular Season. Let's now take a look at pitching & batting stats to see how they impact targetAvg scores.

### Pitching & Batting Stats

There are lots of detailed stats around batting & pitching that are available in the dataset. Let's try to see how good & bad performances lead to engagement.

In [ ]:
player_engagement_with_info.sort_values(by=['playerId','dailyDataDate'],inplace=True)
player_engagement_with_info

dailyDataDate engagementMetricsDate  playerId   target1    target2  \
0          2018-01-01            2018-01-02    112526  0.055277   5.496109   
1          2018-01-02            2018-01-03    112526  0.060625   3.252914   
2          2018-01-03            2018-01-04    112526  0.029341   1.648352   
3          2018-01-04            2018-01-05    112526  0.014799   2.665894   
4          2018-01-05            2018-01-06    112526  0.083916   1.161002   
...               ...                   ...       ...       ...        ...   
2506171    2021-04-26            2021-04-27    685503  0.044617   1.224728   
2506172    2021-04-27            2021-04-28    685503  0.019123   1.178880   
2506173    2021-04-28            2021-04-29    685503  0.015799   4.323489   
2506174    2021-04-29            2021-04-30    685503  0.018770  31.946021   
2506175    2021-04-30            2021-05-01    685503  0.014727   6.607196   

          target3    target4  targetAvg       date  year  ...  \
0        0.025839  16.176471   5.438424 2018-01-01  2018  ...   
1        0.030486   8.541353   2.971345 2018-01-02  2018  ...   
2        0.032613  10.490112   3.050104 2018-01-03  2018  ...   
3        0.087422  19.091467   5.464896 2018-01-04  2018  ...   
4        0.024759   6.643880   1.978389 2018-01-05  2018  ...   
...           ...        ...        ...        ...   ...  ...   
2506171  0.009437   0.737463   0.504062 2021-04-26  2021  ...   
2506172  0.013161   0.790301   0.500366 2021-04-27  2021  ...   
2506173  0.002350   0.970273   1.327977 2021-04-28  2021  ...   
2506174  0.305491   5.938273   9.552139 2021-04-29  2021  ...   
2506175  1.851336   2.014339   2.621900 2021-04-30  2021  ...   

         divisionRankTeam  leagueRankTeam wildCardRankTeam winPctTeam  \
0                     NaN             NaN              NaN        NaN   
1                     NaN             NaN              NaN        NaN   
2                     NaN             NaN              NaN        NaN   
3                     NaN             NaN              NaN        NaN   
4                     NaN             NaN              NaN        NaN   
...                   ...             ...              ...        ...   
2506171               NaN             NaN              NaN        NaN   
2506172               NaN             NaN              NaN        NaN   
2506173               NaN             NaN              NaN        NaN   
2506174               NaN             NaN              NaN        NaN   
2506175               5.0            13.0             10.0      0.407   

        winStreakTeam lossStreakTeam quarter gameday flag daysSinceLastGame  
0                 NaN            NaN       1   False   -1                 1  
1                 NaN            NaN       1   False   -1                 2  
2                 NaN            NaN       1   False   -1                 3  
3                 NaN            NaN       1   False   -1                 4  
4                 NaN            NaN       1   False   -1                 5  
...               ...            ...     ...     ...  ...               ...  
2506171           NaN            NaN       2   False   -1                 1  
2506172           NaN            NaN       2   False   -1                 2  
2506173           NaN            NaN       2   False   -1                 3  
2506174           NaN            NaN       2   False   -1                 4  
2506175           NaN            1.0       2    True    0                 0  

[2506176 rows x 76 columns]

In [ ]:
pitching_stats=player_engagement_with_info.groupby(['pitchingGameScore']).agg({'targetAvg':'mean','playerId':'count'}).reset_index()
pitching_stats

px.scatter(pitching_stats,x='pitchingGameScore',y='targetAvg',title='Variation in targetAvg scores with Tom Tango pitchingGameScore')


In the JSON unpacking code provided by competition organizers, we see them calculating an aggregate pitching statistic, Tom Tango's pitchingGameScore, which gives us an impression of pitcher's performance. Here is how it was calculated -
PitchingGameScore = 40 + 2 * outs + 1 * strikeOutsPitching - 2 * baseOnBallsPitching - 2 * player_game_stats - 3 * runsPitching - 6 * homeRunsPitching

As we can see above, Engagement scores are high for both poor & great pitching performances. However there is a cluster of scores in the 60-85 range where targetAvg values are lower. Since this doesn't show a very clear pattern in targetAvg, maybe we need to individually look at pitcher statistics.

In [ ]:
run_stats=player_engagement_with_info.groupby(['runsScored','homeRuns']).agg({'targetAvg':'mean','target1':'count'}).reset_index()
run_stats['homeRuns']=run_stats['homeRuns'].apply(str)
run_stats.columns=['runsScored','homeRuns','targetAvg','Innings']

team_run_stats=player_engagement_with_info.groupby(['runsScoredTeam']).agg({'targetAvg':'mean','target1':'count'}).reset_index()
team_run_stats.columns=['runsScoredTeam','targetAvg','Innings']
team_run_stats


runsScoredTeam  targetAvg  Innings
0              0.0   2.743258    10282
1              1.0   3.060527    17591
2              2.0   3.329780    22308
3              3.0   3.442415    23627
4              4.0   3.826115    23292
5              5.0   3.946151    21552
6              6.0   4.057941    16439
7              7.0   4.612828    13089
8              8.0   4.457763    10057
9              9.0   4.376931     6998
10            10.0   4.953178     5512
11            11.0   5.115886     4038
12            12.0   5.124909     2436
13            13.0   5.209272     1547
14            14.0   5.673010     1352
15            15.0   5.503103      829
16            16.0   6.451319      528
17            17.0   6.571828      318
18            18.0   5.717769      186
19            19.0   7.361143      205
20            20.0   8.852114       43
21            21.0   6.406145       84
22            22.0   3.202019       45
23            23.0  10.494408       32
24            24.0   2.629656       23
25            25.0  11.582966       14
26            29.0  15.416088       13
27            30.0   7.574099       20

In [ ]:
fig = px.bar(run_stats, x="runsScored", y="targetAvg",
             color='homeRuns', barmode='group',
             height=400)
fig.update_layout(title='Variation of targetAvg with Player runsScored & homeRuns scored in Match')
fig.show()

rbi_stats=player_engagement_with_info.groupby(['rbi']).agg({'targetAvg':'mean'}).reset_index()

fig=px.line(rbi_stats,x='rbi',y='targetAvg',title='Variation in targetAvg values with RunsBattedIn by Player')
fig.show()

fig1 = make_subplots(specs=[[{"secondary_y": True}]])

# Add traces
fig1.add_trace(
    go.Scatter(x=team_run_stats['runsScoredTeam'], y=team_run_stats['targetAvg'], name="targetAvg"),
    secondary_y=True,
)

fig1.add_trace(
    go.Scatter(x=team_run_stats['runsScoredTeam'], y=team_run_stats['Innings'], name="Innings"),
    secondary_y=False,
)
fig1.update_layout(title='Variation of Runs Scored by a Team vs targetAvg')
fig1.show()

Let's now take a look at the impact of batting stats on targetAvg values -
* **General rule seems to be that players who score higher runs in the match tend to have higher engagement**. Baseball is a low scoring game and scoring a run is a big event in the game. Therefore more runs generally means more engagment.
* **Home Runs are also a rare event in the sport of Baseball. Greater the number of Home Runs scored in the game, higher the engagement.**
* Another statistic for batter performance is RBI - Runs Batted In. It shows that the number of runs scored by a team where a player was involved in. This also shows that higher the RBI number, greater the engagement for a player
* Also the **team as a whole tends to have higher engagement levels when the runsScored are higher**.

Since this is essentially a time series dataset, we should also look at autocorrelations for target variables - autocorrelation measures correlation of a time series value at any point with its previous values. For every player, we will calculate autocorrelation at various lags and look at the distribution of autocorrelation for different lag values.

In [ ]:
out_stats=player_engagement_with_info.groupby(['strikeOuts']).agg({'targetAvg':'mean','target1':'count'}).reset_index()
out_stats.columns=['strikeOuts','targetAvg','Num_Days']
#px.line(out_stats,x='strikeOuts',y='targetAvg')

fig = make_subplots(specs=[[{"secondary_y": True}]])

# Add traces
fig.add_trace(
    go.Scatter(x=out_stats['strikeOuts'], y=out_stats['targetAvg'], name="targetAvg"),
    secondary_y=True,
)

fig.add_trace(
    go.Scatter(x=out_stats['strikeOuts'], y=out_stats['Num_Days'], name="Num_Days"),
    secondary_y=False,
)
fig.update_layout(xaxis_title='Count of Strike Outs',title='StrikeOuts in a Game vs Target Score')
fig.show()

pitch_run_stats=player_engagement_with_info.groupby(['runsPitching']).agg({'targetAvg':'mean','target1':'count'}).reset_index()
pitch_run_stats.columns=['runsPitching','targetAvg','Num_Days']

fig = make_subplots(specs=[[{"secondary_y": True}]])

# Add traces
fig.add_trace(
    go.Scatter(x=pitch_run_stats['runsPitching'], y=pitch_run_stats['targetAvg'], name="targetAvg"),
    secondary_y=True,
)

fig.add_trace(
    go.Scatter(x=pitch_run_stats['runsPitching'], y=pitch_run_stats['Num_Days'], name="Num_Days"),
    secondary_y=False,
)
fig.update_layout(xaxis_title='Count of Runs Allowed',title='Runs Allowed in a Game vs targetAvg')
fig.show()


In [ ]:
### group by year?
player_avg = player_engagement_with_info.groupby(["playerId"])[["target1","target2","target3","target4"]].mean().reset_index()
player_median = player_engagement_with_info.groupby(["playerId"])[["target1","target2","target3","target4"]].median().reset_index()
player_mean=pd.concat([player_avg, player_median], axis=1).groupby(axis=1, level=0).mean()
player_mean["target1"] = .85 * player_mean["target1"]
player_mean["target2"] = .85 * player_mean["target2"]
player_mean["target3"] = .85 * player_mean["target3"]
player_mean["target4"] = .85 * player_mean["target4"]
gc.collect()
print(player_mean.shape)
player_mean

(2061, 5)


playerId   target1   target2   target3    target4
0     112526.0  0.423243  2.835345  0.989628  10.971118
1     134181.0  0.798428  1.990238  1.190390   4.977005
2     279571.0  0.004454  0.192152  0.020792   0.201132
3     282332.0  1.566413  4.516149  2.204880   9.003971
4     400085.0  0.836112  3.898183  1.464149  16.296514
...        ...       ...       ...       ...        ...
2056  681911.0  0.001395  0.245447  0.163320   0.040023
2057  683232.0  0.000282  0.099019  0.005570   0.009447
2058  683734.0  0.222503  0.546168  0.573791   0.612408
2059  685493.0  0.027261  0.993570  0.044770   0.354285
2060  685503.0  0.015479  0.313719  0.017962   0.120708

[2061 rows x 5 columns]

In [ ]:
player_engagement_with_info.columns

Index(['dailyDataDate', 'engagementMetricsDate', 'playerId', 'target1',
       'target2', 'target3', 'target4', 'targetAvg', 'date', 'year', 'month',
       'inSeason', 'seasonPart', 'playerName', 'DOB', 'mlbDebutDate',
       'birthCity', 'birthStateProvince', 'birthCountry',
       'primaryPositionName', 'rosterStatusCode', 'rosterStatus',
       'rosterTeamId', 'rosterTeamName', 'numGames', 'numTeams', 'gameTeamId',
       'runsScored', 'homeRuns', 'strikeOuts', 'baseOnBalls', 'hits',
       'hitByPitch', 'atBats', 'caughtStealing', 'stolenBases',
       'groundIntoDoublePlay', 'groundIntoTriplePlay', 'plateAppearances',
       'totalBases', 'rbi', 'leftOnBase', 'sacBunts', 'sacFlies',
       'gamesStartedPitching', 'runsPitching', 'homeRunsPitching',
       'strikeOutsPitching', 'baseOnBallsPitching', 'hitsPitching',
       'inningsPitchedAsFrac', 'earnedRuns', 'battersFaced', 'saves',
       'blownSaves', 'pitchingGameScore', 'noHitter', 'gameTeamName',
       'gameType', 'oppId',

In [ ]:
data=player_engagement_with_info[['dailyDataDate','playerId','target1','target2','target3','target4','targetAvg','quarter','gameday']]

In [ ]:
autocorrel_list=list()
for i in range(31):
    ser=data.groupby('playerId')['targetAvg'].apply(lambda x: x.autocorr(lag=i))
    df=ser.to_frame().reset_index()
    df['lag']=i
    autocorrel_list.append(df)
auto_frame=pd.concat(autocorrel_list).reset_index(drop=True)

autocorrel_list

[      playerId  targetAvg  lag
 0       112526        1.0    0
 1       134181        1.0    0
 2       279571        1.0    0
 3       282332        1.0    0
 4       400085        1.0    0
 ...        ...        ...  ...
 2056    681911        1.0    0
 2057    683232        1.0    0
 2058    683734        1.0    0
 2059    685493        1.0    0
 2060    685503        1.0    0
 
 [2061 rows x 3 columns],
       playerId  targetAvg  lag
 0       112526   0.540552    1
 1       134181   0.657213    1
 2       279571   0.457651    1
 3       282332   0.493433    1
 4       400085   0.734707    1
 ...        ...        ...  ...
 2056    681911   0.408333    1
 2057    683232   0.367140    1
 2058    683734   0.670498    1
 2059    685493   0.601237    1
 2060    685503   0.560925    1
 
 [2061 rows x 3 columns],
       playerId  targetAvg  lag
 0       112526   0.254302    2
 1       134181   0.435061    2
 2       279571   0.333342    2
 3       282332   0.131600    2
 4       400085 

In statistics and time series analysis, "lag" refers to the position of a data point relative to its position in time. A lag of 1 means the data point is observed one time unit earlier, a lag of 2 is two time units earlier, and so on. In autocorrelation plots, these lags are used to understand how a data series is correlated with itself over different time intervals. For example, in your autocorrelation plots for targetAvg, a lag of 1 compares each data point with its immediate predecessor, while a lag of 30 compares each point with the one 30 time units before. This helps in identifying patterns and trends that occur over time in the data.


This script creates a grid of histograms (5 rows and 6 columns) where each histogram represents the distribution of targetAvg for different lag values (from 1 to 30). The make_subplots function is used to create the subplot structure, and for each lag, a histogram is added to the appropriate subplot position. The figure layout is then updated, and the plot is displayed.

In [ ]:
# Import required libraries
import plotly.graph_objects as go
from plotly.subplots import make_subplots

# Create a list of titles for the subplots
lst = []
for i in range(1, 31):
    lst.append('lag_' + str(i))
tup = tuple(lst)

# Initialize a figure with 5 rows and 6 columns of subplots
fig = make_subplots(rows=5, cols=6, shared_yaxes=True, subplot_titles=tup)
k = 1
j = 1

# Loop through 30 lags to create a histogram for each
for i in range(1, 31):
    # Filter the frame for each lag
    lag_frame = auto_frame.loc[auto_frame['lag'] == i]

    # Add a histogram trace for each lag
    fig.add_trace(
        go.Histogram(x=lag_frame['targetAvg'], name=str(i)),
        row=k, col=j
    )

    # Update column index, and if it's at the end, move to next row
    j = j % 6
    j = j + 1
    if (i % 6 == 0):
        k = k + 1

# Update the layout of the figure and display it
fig.update_layout(height=1000, title_text="Autocorrelation Plots for player targetAvg for various lags", showlegend=False)
fig.show()


In [ ]:
lag_gp=auto_frame.groupby('lag').agg({'targetAvg':['mean','median']}).reset_index()
lag_gp.columns=['lag','mean','median']
fig=go.Figure()
fig.add_trace(go.Scatter(x=lag_gp['lag'],y=lag_gp['mean'],mode='lines',name='mean'))
fig.add_trace(go.Scatter(x=lag_gp['lag'],y=lag_gp['median'],mode='lines',name='median'))
fig.update_layout(title='Mean & Median of player Autocorrelation distributions across various lag periods',xaxis_title='lag')
